File: Алгогритмы ПР2.cpp
1.Бинарная куча (Binary Heap)
class MinHeap {
    vector<int> heap;
    void heapifyUp(int index) {
        while (index > 0 && heap[index] < heap[(index-1)/2]) {
            swap(heap[index], heap[(index-1)/2]);
            index = (index-1)/2;
        }
    }
};
2.Хеш-таблицы (Hash Tables)
#include <unordered_map>
std::unordered_map<std::string, //int> map;#std::unordered_map предоставляет типобезопасную реализацию с возможностью кастомных хеш-функций и политик управления памятью.
map["key"] = 42;  // Шаблонная реализация
3. Числа Фибоначчи и алгоритмы
int fibonacci(int n) {
    if (n <= 1) return n;
    int a = 0, b = 1;
    for (int i = 2; i <= n; i++) {
        int temp = a + b;
        a = b;
        b = temp;
    }
    return b;  // O(n) время, O(1) память
}

File: Алгогритмы ПР2.java
1.Бинарная куча (Binary Heap)
public class BinaryHeap<T extends Comparable<T>> {
    private ArrayList<T> heap;
    // Использует Comparable для сравнения элементов
}
2.Хеш-таблицы (Hash Tables)
HashMap<String, Integer> map = new HashMap<>();
map.put("key", 42);  // Автоматическое boxing/unboxing
3. Числа Фибоначчи и алгоритмы
public static long fibIterative(int n) {
 // Обрабатываем базовые случаи 
if (n <= 1) {
 return n;
 } 
// Инициализируем два первых числа последовательности 
long a = 0; // F(0) 
long b = 1; // F(1)
 // Цикл от 2 до n включительно for (int i = 2; i <= n; i++) {
 // Вычисляем следующее число как сумму двух предыдущих
 long next = a + b; 
// Сдвигаем переменные: a становится b, b становится next 
a = b; // Теперь a = F(i-1) 
b = next; // Теперь b = F(i) 
} 
// Возвращаем n-е число Фибоначчи 
return b;
 }